Extraneous grass block generation in certain biomes when spawning meteors

Gamingb3ast opened the pull request:
I've often noticed that whenever meteors generate in other dimensions or in specific biomes, they will generate with the grass as usual. This can look odd in some cases and, albeit not very major, be a bit unpleasant.

This one simple if statement will, before attempting to generate grass, check if the biome's top block (in BiomeGenBase) is Blocks.grass. If it is, then the generation continues as normal. If it's not, then it will instead generate dirt, which is not as out of place in nearly all cases. It may be a good idea to add a config for the grass substitution in cases where dirt is out of place (it could instead, for example, be air or lava); however, that's something that I felt should be discussed with you before it is implemented. It may also be a good idea to have it generate the biome top block instead of dirt, which will guarantee that in 99% of cases, it looks correct.

I've done testing with the wasteland mod biomes and stone shores/peaks. I believe this will not cause any issues as the simplicity of the code and its use of the most basic biome top block detection will insure correct usage.

This will not have any effect on performance, and I feel will not be an addition that is disliked. I hope to see it added, and I understand if it isn't.

TL:DR

Grass will now only generate if the meteor generated in a biome where grass is the top block. 

This fixes some of the odd-looking meteors when they spawn in certain biomes or dimensions.

Gamingb3ast commented:
I made slight edits that talk about other alternatives to which block substitutes grass. Currently it is dirt, as it already generates around meteors and is less out of place in nearly all cases were grass is out of place. The edit adds to the idea of the config by instead offering another suggestion, which would be to have it generate the top block of the biome. It could even be optimized by doing something similar to 

```java
this.putter.put(w, x, y, z, w.getWorld().getBiomeGenForCoords(x, z).topBlock);
```

which would completely remove the need for the if statement and look perfect nearly every time. However this change would be a bit larger than just substituting grass for dirt in a few rare cases, which is why the initial pull request does not include it.

LazyFleshWasTaken commented:
Why not directly use the detected non-grass block instead of using dirt? This would eliminate the extra times where dirt looks weird too.

Gamingb3ast commented:
> Why not directly use the detected non-grass block instead of using dirt? This would eliminate the extra times where dirt looks weird too.
> 
> edit: too late

Yeah, I just haven't worked with GTNH forks of mods before, so I was a bit cautious with my changes. 

About the extra times dirt would look odd, it anyways generates with or without the grass (The debris that generates normally is: Stone, cobblestone, dirt, gravel, grass, and skystone.) So I chose to use dirt, however, I absolutely agree that using the detected block would be better! 

In the case of dirt, I'm unsure of whether any changes should be made as in cases where grass doesn't generate, most of the time it's some alternative version of soil. (For instance, Galaxy Space and other GC addons use their own grasses instead of recoloring the existing grass using biome colors.)

Gamingb3ast commented:
I've modified the code to comply with the formatting guidelines. I've also changed it to use the top and filler blocks instead of a grass check, which means it will work far better in more biomes. 

I'm not sure whether I need to make a new PR or if it will be retested. 

Here are some test screenshots I took:

![2025-04-04_18 31 41](https://github.com/user-attachments/assets/6b9a7831-e489-4724-9dc8-46b54633ef48)
![2025-04-05_12 08 05](https://github.com/user-attachments/assets/0b25c82b-2e32-4bf6-878a-acbaa3161435)
![2025-04-05_12 07 44](https://github.com/user-attachments/assets/70647ab8-b6e0-4ef6-bb34-74b99bbf462f)
![2025-04-05_12 08 18](https://github.com/user-attachments/assets/5dd63562-d143-4aef-9942-60191dbff874)
![2025-04-05_12 33 45](https://github.com/user-attachments/assets/b3df32c3-80a6-4657-acd1-70f6f96a342d)

serenibyss reviewed (approved):
Nice fix, thank you!

Gamingb3ast commented:
> Nice fix, thank you!

Happy to do my part! ❤️

AnrDaemon commented:
@Gamingb3ast sorry for reaching you like this, can you please ping or DM me on discord? I'd like to talk a question or two regarding AE2 worldgen.

Gamingb3ast commented:
> @Gamingb3ast sorry for reaching you like this, can you please ping or DM me on discord? I'd like to talk a question or two regarding AE2 worldgen.

Yeah what's your username?
